File: projects/IST-IG-DD-JAVA/transforms-java/src/main/java/myproject/datasets/exposure/typed/fp_power_m01015.java
package myproject.datasets.exposure.typed;


import com.palantir.transforms.lang.java.api.*;

import com.palantir.transforms.lang.java.api.Compute;
import com.palantir.transforms.lang.java.api.Input;
import com.palantir.transforms.lang.java.api.Output;
import com.palantir.transforms.lang.java.api.FoundryOutput;
import com.palantir.transforms.lang.java.api.FoundryInput;
import com.palantir.transforms.lang.java.api.ReadRange;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.functions;
import myproject.datasets.util.BP_ReadRangeUtils;
import myproject.datasets.util.BP_WriteRangeUtils;

/**
 * This is an example high-level Transform intended for automatic registration.
 */
public final class fp_power_m01015 {

    // The class for an automatically registered Transform contains the compute
    // function and information about the input/output datasets.
    // Automatic registration requires "@Input" and "@Output" annotations.
    @Compute      
    public void myComputeFunction(@Input("/BP/IST-IG-SS-Systems/data/raw/freeport/fp_power_m01015/fp_power_m01015") FoundryInput input,
    @Output("/BP/IST-IG-DD/data/technical/exposure_java/typed/freeport/fp_power_m01015") FoundryOutput output
    ) {
        ReadRange readRange = BP_ReadRangeUtils.getReadRange(input);
        Dataset<Row> df = input.asDataFrame().readForRange(readRange);

        df = df.na().drop("all", new String[] {"Valuation_Price"});
        df = df.select(
                functions.to_date(df.col("PLEX_Calculation_Date"), "yyyy/MM/dd").alias("PLEX_Calculation_Date"),
                df.col("Team_Name"),
                df.col("Book_Name"),
                df.col("Counterparty_Name"),
                df.col("Curve_Short_Name"),
                df.col("Notional_Qty_-_Current").cast("double").alias("Notional_Qty_Current"),
                functions.to_date(df.col("Trade_Date"), "yyyy/MM/dd").alias("Trade_Date"),
                df.col("Trade_ID").cast("long").alias("Trade_ID"),
                df.col("Commodity_Name"),
                df.col("Trade_Type_Code"),
                functions.to_date(df.col("Delivery_From_Period_Summary")).alias("Delivery_From_Period_Summary"),
                df.col("Valuation_Price").cast("double"),
                df.col("MTM_Price_-_Current").cast("double").alias("MTM_Price_Current"),
                df.col("MTM_Amt_-_Current").cast("double").alias("MTM_Amt_Current"),
                df.col("Disc_MTM_Amt_-_Current").cast("double").alias("Disc_MTM_Amt_Current"),
                df.col("Hedge_Qty_-_Current").cast("double").alias("Hedge_Qty_Current"),
                functions.to_timestamp(df.col("Run_Datetime"), "yyyy-MM-dd HH:mm:ss").alias("Run_Datetime")
        );
        output.getDataFrameWriter(df).write(BP_WriteRangeUtils.getWriteRange(input));
    }
}

File: projects/integration/transforms-python/src/python/exposure/typed/fp_power_m01015.py
from transforms.api import transform_df, incremental, Input, Output
import pyspark.sql.functions as F


@incremental()
@transform_df(
    Output("/BP/IST-IG-DD/data/technical/exposure/typed/freeport/fp_power_m01015"),
    df=Input("/BP/IST-IG-SS-Systems/data/raw/freeport/fp_power_m01015/fp_power_m01015"),
)
def my_compute_function(df):
    # TODO: Reconsider how we handle this case
    # We have empty rows in raw file - fillna to prevent error on cast
    # df = df.fillna(0.0, subset=["Valuation_Price"])

    # df = df.na.drop(thresh=len(df.columns)) # garbage rows are blank in all cols except Disc_MTM_Amt_Current
    df = df.na.drop(how='all', subset=["Valuation_Price"])
    df = df.select(
        F.to_date(df.PLEX_Calculation_Date, 'yyyy/MM/dd').alias('PLEX_Calculation_Date'),
        df.Team_Name,
        df.Book_Name,
        df.Counterparty_Name,
        df.Curve_Short_Name,
        df['Notional_Qty_-_Current'].cast('double').alias('Notional_Qty_Current'),
        F.to_date(df.Trade_Date).alias('Trade_Date'),
        df.Trade_ID.cast('long').alias('Trade_ID'),
        df.Commodity_Name,
        df.Trade_Type_Code,
        F.to_date(df.Delivery_From_Period_Summary).alias('Delivery_From_Period_Summary'),
        df.Valuation_Price.cast('double'),
        df['MTM_Price_-_Current'].cast('double').alias('MTM_Price_Current'),
        df['MTM_Amt_-_Current'].cast('double').alias('MTM_Amt_Current'),
        df['Disc_MTM_Amt_-_Current'].cast('double').alias('Disc_MTM_Amt_Current'),
        df['Hedge_Qty_-_Current'].cast('double').alias('Hedge_Qty_Current'),
        F.to_timestamp(df['Run_Datetime'], 'yyyy-MM-dd HH:mm:ss').alias('Run_Datetime'),
    )
    return df
